### DISCUSSION TITLE
Auto Force Syncing

### DISCUSSION THREAD

**Discussion by rumstead** [argoproj/argo-cd#5172] [Q&A] [answered]
Hi all! I want to be able to automatically “kubectl apply --force” my Kubernetes manifests so I don't have to force sync via the UI when a label or something immutable changes. I saw the [SyncOptions](https://github.com/argoproj/argo-cd/blob/v1.8.1/pkg/apis/application/v1alpha1/types.go#L590) and tried adding “--force”, “force=true”, and “Force=true” but none seemed to work. I then stumbled across [SyncStrategy](https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L735). I was able to set this on the Application Resource and it seems to work but I want to make sure this is the best way to do it.

**Comment by sbose78** [comment, marked as answer]
Hi @rumstead ,

`force=true` as a sync option is not supported.

The following are the only valid Sync Options:

* Prune=false
* Validate=false
* SkipDryRunOnMissingResource=true

> I then stumbled across SyncStrategy. I was able to set this on the Application Resource and it seems to work but I want to make sure this is the best way to do it.

You are right.

`.spec.operation.sync.syncStrategy.apply.force` needs to be set to `true` . This would delete and re-create the resource, if PATCH encounters a conflict and has retried for 5 times.

**Reply by jessesuen** [reply]
Be extremely careful about force syncing an application which contains CRDs. If the force sync is needed on a CRD, it will mean that all CR instances will be deleted as well. For example, recently a change was made to cert-manager CRDs that caused it to not be updatable through normal `kubectl apply`. Force syncing of the cert-manager application caused the CRDs to be deleted, and subsequently *all CR instances* (Certificates, Issuers) to be deleted as well.

**Reply by nichcuta** [reply]
Can someone kindly provide a full example for `.spec.operation.sync.syncStrategy.apply.force` or link me to the relevant docs? Cannot seem to get it working or find related docs.

**Reply by gitromba** [reply]
@nichcuta there's none for the time being. I ended up here after bit of googling.

**Reply by gitromba** [reply]
For future reference: with ArgoCD `2.0.3` this is on the same level with `spec` making it `.operation.sync.syncStrategy.apply.force`.
Unfortunately when creating a new Argo App with this attributes one will experience no effect. Seems like gets overwritten by ArgoCD and used internally to manage state: https://github.com/argoproj/argo-cd/blob/1b46143d0777698dd0e9751e80bd713aeb79a72f/manifests/install.yaml#L50

**Reply by nichcuta** [reply]
What I ended up doing for the time being is, creating the argocd app without syncStrategy defined. After a successful creation of app, I added status.operationState.operation.sync.syncStrategy.apply.force https://github.com/argoproj/argo-cd/blob/master/manifests/crds/application-crd.yaml#L402 which works.

Example Yaml Def:
`spec:.....
status:
  operationState:
    operation:
      sync:
        syncStrategy:
          apply:
            force: true`

Note that I would consider this a temporary hack to get the --force working since you are editing the state of the app and this cannot be done on app creation since it would fail due to missing mandatory parameters.

Hope this helps someone

**Comment by Demonsthere** [comment]
Hi @sbose78 are you sure `SkipDryRunOnMissingResource` is still valid? I'm running Argo 2.0 and couldn't get this option to work

**Comment by awelsman** [comment]
Hey, is this workaround still relevant? I am using 2.2.3 and cannot seem to get it to work when applying `.spec.operation.sync.syncStrategy.apply.force` or `.operation.sync.syncStrategy.apply.force` to true (or any other examples listed here) on an Application post creation to always run a force apply during a sync.

We are moving away from simple automated fire and forget `helm upgrade --force` commands and utilizing Argo, but the team has gotten used to the fact that a force apply occurs every single time.

Thanks!

**Comment by ghost** [comment]
The suggested fix works for me, defining `.operation.sync.syncStrategy.apply.force` using ArgoCD 2.4.2.

But, using another ArgoCD Application to deploy this one, the operation part of the manifest is stripped away by ArgoCD. This renders the application always out of sync. Is there a solution for this behaviour?

![grafik](https://user-images.githubusercontent.com/96055817/202241370-bcdb964b-6dd7-4bf5-94e9-72d5a37d801e.png)

**Reply by pierreyves-lebrun** [reply]
> The suggested fix works for me, defining `.operation.sync.syncStrategy.apply.force` using ArgoCD 2.4.2.
> 
> But, using another ArgoCD Application to deploy this one, the operation part of the manifest is stripped away by ArgoCD. This renders the application always out of sync. Is there a solution for this behaviour?
> 
> ![grafik](https://user-images.githubusercontent.com/96055817/202241370-bcdb964b-6dd7-4bf5-94e9-72d5a37d801e.png)

Adding the following fields to the parent app worked for me:
```
spec:
  ignoreDifferences:
    - group: [argoproj.io](http://argoproj.io/)
      kind: Application
      jsonPointers:
        - /operation
```

**Reply by ghost** [reply]
This works like a charm, thank you!
